### ISSUE TITLE
i686-linux bash segfaults in QEMU VM on x86_64 host

### ISSUE THREAD

**Issue by yu-re-ka** [NixOS/nixpkgs#133608]
## Edit: 
Wow this is much more complicated (to reproduce). Will probably have to open a bug in QEMU or something.

### Steps To Reproduce

1. Create a QEMU VM with CPU model host-passthru on a Ryzen host
1. In the VM: `nix-build "<nixpkgs>" -A bash --system i686-linux`
2. In the VM: `result/bin/bash`
3. If it segfaulted: yay, you reproduced it, if not: close the shell, goto step 2

### Expected behavior

It should work reliably on all i686 and x86_64 CPUs

### Additional context

I found this while building robotnix: https://github.com/danielfullmer/robotnix/issues/109
It builds something like this:

```
with import <nixpkgs> {};

buildFHSUserEnv {
  name = "chromium-fhs";
  multiPkgs = pkgs: with pkgs; [
    libgcc
  ];
}
```

Which will fail on my systems, because the bash used by i686 stdenv in the libgcc build segfaults:

```
builder for '/nix/store/6d4sagmni5znazlis98vif102s2mxm2r-stdenv-linux.drv' failed due to signal 11 (Segmentation fault)
cannot build derivation '/nix/store/kpaa8jvydclzxbhr8n07a5wq6fbc2kls-libgcc-10.3.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/zm1bvxshp6wmc5983wzfyrhs5ahwnnay-chromium-fhs-usr-multi.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/36pbxikhh8x9p5sp10pzhcbvp4rs4csl-chromium-fhs-fhs.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/fhkxy0617kl2rcndwv79gaax2swgfgig-chromium-fhs-init.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/c2cf73glh6vcqlbg0jgfvv07gzi9qx06-chromium-fhs.drv': 1 dependencies couldn't be built
error: build of '/nix/store/c2cf73glh6vcqlbg0jgfvv07gzi9qx06-chromium-fhs.drv' failed
``` 

### Notify maintainers

@peti @dtzWill 

### Metadata

nixpkgs 9924d9231d07d8610b866b9011cf1d02fd4df616
AMD Ryzen 9 3950X

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: bash
# a list of nixos modules affected by the problem
module:
```

**Comment by yayayayaka** [comment]
>     2. `nix-build "<nixpkgs>" -A bash --system i686-linux`
> 
>     3. In the VM: `result/bin/bash`

Cannot reproduce.

**Comment by yu-re-ka** [comment]
@danielfullmer can you confirm this with your Threadripper 3970X?

```
$ wget https://channels.nixos.org/nixos-21.05/latest-nixos-minimal-x86_64-linux.iso
$ qemu-system-x86_64 -M pc,accel=kvm -m 2G -cpu EPYC-Rome -cdrom latest-nixos-minimal-x86_64-linux.iso
```
(`-cpu host` or `-cpu EPYC-Rome` doesn't seem to matter)

In the VM:

```
$ nix-build "<nixpkgs>" -A pkgsi686Linux.bash
$ result/bin/bash
```

**Comment by OPNA2608** [comment]
The above example seems to work fine on my Threadripper 2970X, 21.05 host in case it matters.

**Comment by yu-re-ka** [comment]
I suspect it might only happen on Zen 2 hosts

**Comment by yu-re-ka** [comment]
Okay, thanks everyone. I will try to get my hands on a second Ryzen (not Threadripper) 3000 CPU and test there, but it might take a bit. If this is really only on my one machine I guess it might be a hardware fault, but I doubt that.

**Comment by danielfullmer** [comment]
@yu-re-ka Just to confirm, you've encountered this issue on both "AMD Ryzen 9 3950X" as well as "AMD Ryzen 7 Pro 4750G" ?

**Comment by yu-re-ka** [comment]
I thought I had the issue on the Ryzen 7 Pro 4750G months ago, but I could only reproduce on the Ryzen 9 3950X this time.

**Comment by vcunat** [comment]
I suspect it's the same issue as what's killing almost all i686 NixOS tests on Hydra: https://hydra.nixos.org/eval/1709986?filter=i686-linux&compare=release-21.05#tabs-still-succeed

I can reproduce the test failures locally, but apparently for me they started later in the `release-21.05..master` range than on Hydra.  Perhaps there's an impurity involved, e.g. the around the host's kernel.

**Comment by yu-re-ka** [comment]
Indeed! Downgrading the host's kernel to 5.10.70 (previously 5.14.x) helped. So one factor is the host kernel version.

**Comment by vcunat** [comment]
I'm using NixOS host with 5.10.69 now, and reverting glibc 2.32 -> 2.33 fixed an i686 test.  That revert was _both_ for qemu and the test itself.  Maybe the first was the key; qemu even needed patching to build after glibc update (04be5ced7e).

**Comment by jkarlson** [comment]
We had an issue that we believe to be similar, other notes:

- `qemu-system-i386` does complain on AMD cpus the cpu is not smp safe, we did not ascertain, whether this false positive is correlated to any issue, only that it does not happen in intel cpus.
- We tried `-cpu host` to see, if it would affect any of our issues, this only caused other issues (need to check what, if it becomes relevant).
- using `qemu-system-x86_64` with 32-bit kernel and userland made the issue go away.
